### PR TITLE
[WIP] general fix to catkin build

### PR DESCRIPTION
**Pull request by TSC21**
Fixes https://github.com/PX4/Firmware/issues/9447, https://github.com/PX4/Firmware/issues/9554, https://github.com/PX4/Firmware/issues/10089, https://github.com/PX4/Firmware/issues/10393.

It builds but currently failing on linking `perf` lib with the `logger` module.

### CODE REVIEW

**Comment by TSC21** [comment]
```bash
nuno@nuno-P95xER:~/PX4/Firmware_dev$ catkin build
----------------------------------------------------------------
Profile:                     default
Extending:             [env] /opt/ros/melodic
Workspace:                   /home/nuno/PX4/Firmware_dev
----------------------------------------------------------------
Source Space:       [exists] /home/nuno/PX4/Firmware_dev/src
Log Space:          [exists] /home/nuno/PX4/Firmware_dev/logs
Build Space:        [exists] /home/nuno/PX4/Firmware_dev/build
Devel Space:        [exists] /home/nuno/PX4/Firmware_dev/devel
Install Space:      [unused] /home/nuno/PX4/Firmware_dev/install
DESTDIR:            [unused] None
----------------------------------------------------------------
Devel Space Layout:          linked
Install Space Layout:        None
----------------------------------------------------------------
Additional CMake Args:       None
Additional Make Args:        None
Additional catkin Make Args: None
Internal Make Job Server:    True
Cache Job Environments:      False
----------------------------------------------------------------
Whitelisted Packages:        None
Blacklisted Packages:        None
----------------------------------------------------------------
Workspace configuration appears valid.
----------------------------------------------------------------
[build] Found '1' packages in 0.0 seconds.                                                                                             
[build] Package table is up to date.                                                                                                   
Warning: generated devel space setup files have been deleted.
Starting  >>> catkin_tools_prebuild                                                                                                    
Finished  <<< catkin_tools_prebuild                [ 0.6 seconds ]                                                                     
Starting  >>> px4                                                                                                                      
_______________________________________________________________________________________________________________________________________
Errors     << px4:make /home/nuno/PX4/Firmware_dev/logs/px4/build.make.000.log                                                         
/usr/bin/x86_64-linux-gnu-ld: /home/nuno/PX4/Firmware_dev/devel/.private/px4/lib/libmodules__logger.a(logger.cpp.o): undefined reference to symbol 'perf_iterate_all'
//home/nuno/PX4/Firmware_dev/devel/.private/px4/lib/libperf.so: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [bin/px4] Error 1
make[1]: *** [platforms/posix/CMakeFiles/px4.dir/all] Error 2
make: *** [all] Error 2
cd /home/nuno/PX4/Firmware_dev/build/px4; catkin build --get-env px4 | catkin env -si  /usr/bin/make --jobserver-fds=6,7 -j; cd -
.......................................................................................................................................
Failed     << px4:make                             [ Exited with code 2 ]                                                              
Failed    <<< px4                                  [ 24.7 seconds ]                                                                    
[build] Summary: 1 of 2 packages succeeded.                                                                                            
[build]   Ignored:   None.                                                                                                             
[build]   Warnings:  None.                                                                                                             
[build]   Abandoned: None.                                                                                                             
[build]   Failed:    1 packages failed.                                                                                                
[build] Runtime: 25.3 seconds total.                                                                                                   
[build] Note: Workspace packages have changed, please re-source setup files to use them.
```
It almost finishes, but crashes on the symlink error. To reproduce, create a catkin ws, create a `src` dir inside and put the Firmware folder inside. Then, on the catkin ws root, issue `catkin build` (you should have `python-catkin-tools` installed).

**Comment by Stifael** [comment]
I am not sure if that is the right fix because you haven't changed the CMakeLists.txt files. There must be something else wrong.
All the individual FlightTasks CMakeLists.txt are linking to the appropriate dependencies. Once this is done, the compiler should be able to link all individual libraries correctly such that you only have to add the header without the any additional file-paths. 
On the other hand, I do not know why the it fails when using catkin. It seems the build system is set up differently for catkin

**Comment by TSC21** [comment]
> On the other hand, I do not know why the it fails when using catkin. It seems the build system is set up differently for catkin

@Stifael it is. That's why this add to be done differently to support ECL and FlightTasks build. A reminder that are many project that include PX4 Firmware inside a catkin workspace, so that's why I am bringing this in order to fix it.

**Comment by Stifael** [comment]
So is this a temporary fix or the actual solution?

**Comment by TSC21** [comment]
> I am not sure if that is the right fix because you haven't changed the CMakeLists.txt files.

I didn't directly changed them but I did changed the `px4_base.cmake` module, which directly influences the build.

**Comment by TSC21** [comment]
> So is this a temporary fix or the actual solution?

That I can't currently tell. Whatever fixes I tried to implement on the `CMakeLists.txt` of FlightTasks resulted in the same issues. This is probably because most of the includes and dependencies are target directed, while catkin probably expects some general includes.

**Comment by Stifael** [comment]
> I didn't directly changed them but I did changed the px4_base.cmake module, which directly influences the build.

Sure, I saw that. But the way I understood it was the PX4 cmake structure should become more modular. For instance, the library `FlightTasks` is already added by the modules who are using it (mc_pos_control_main: https://github.com/PX4/Firmware/blob/master/src/modules/mc_pos_control/CMakeLists.txt#L44). If you just add all the libraries to px4_base, then the modular way is somewhat bypassed

**Comment by TSC21** [comment]
> If you just add all the libraries to px4_base, then the modular way is somewhat bypassed

I know that, but I couldn't find a way of actually have catkin work with making the library visible.

**Comment by Stifael** [comment]
no, not yet. I was just wondering if you are aware of that. 
Basically with this PR, all the linking within the different tasks are completely useless https://github.com/PX4/Firmware/blob/master/src/lib/FlightTasks/tasks/Auto/CMakeLists.txt#L38-L39. That just had to be known when merging this PR. I even would then suggest to remove these lines entirely, because they do not do anything

File: cmake/common/px4_base.cmake [hunk 1]
```diff
@@ -511,6 +511,7 @@ function(px4_add_common_flags)
 		${PX4_BINARY_DIR}/src
 		${PX4_BINARY_DIR}/src/lib
 		${PX4_BINARY_DIR}/src/modules
+		${PX4_BINARY_DIR}/src/lib/FlightTasks
```
Comment (dagar) [inline cmake/common/px4_base.cmake]: Add comments that these are a temporary workaround and need to be fixed. PX4 builds without FlightTasks, ecl, matrix should be entirely possible.

**Comment by dagar** [comment]
@TSC21 try adding perf as a logger dependency (DEPENDS). https://github.com/PX4/Firmware/blob/master/src/modules/logger/CMakeLists.txt#L46

**Comment by dagar** [comment]
Could you add the simple catkin CI build to this branch as well? https://github.com/PX4/Firmware/pull/9679
Then we'd know if everything is actually working together.

**Comment by TSC21** [comment]
> @TSC21 try adding perf as a logger dependency (DEPENDS). https://github.com/PX4/Firmware/blob/master/src/modules/logger/CMakeLists.txt#L46

Already tried that with the same effect.

> Could you add the simple catkin CI build to this branch as well? #9679
Then we'd know if everything is actually working together.

This currently isn't working but I guess that would help debug it.

File: cmake/common/px4_base.cmake (hunk 1 above)
Comment (TSC21) [reply cmake/common/px4_base.cmake]: I will do that but just wanted to have the build fixed before tight things.

**Comment by TSC21** [comment]
Superseded by #10587

**Pull request closed without merging**